### PR TITLE
Use TopologicalSort::minSort to order rec groups

### PR DESCRIPTION
**Pull request by tlively**
Rec groups need to be topologically sorted for the output module to be
valid, but the specific order of rec groups also affects the module size
because types at lower indices requires fewer bytes to reference. We
previously optimized for code size when gathering types by sorting the
list of groups before doing the topological sort. This was brittle,
though, and depended on implementation details of the topological sort
to be correct.

Replace the old topological sort with use of the new
`TopologicalSort::minSort` utility, which is a more principled method of
achieving a minimal topological sort with respect to some comparator.
Also draw inspiration from ReorderGlobals and apply an exponential
factor to take the users of a rec group into account when determining
its weight.

### CODE REVIEW

**Comment by tlively** [comment]
Here's the result of the experiment I did to find the best childFactor for calculating weights. The X-axis is the various factors from 0 to 1. The Y-axis is the number of bytes taken up by all type indices in the module as a proportion of the minimum observed index bytes for each module. (The effects on total module size are much smaller.)

chart

It's interesting that the different modules have such different characteristics, but any factor outside the range .1-.6 is clearly bad across the board.

Unfortunately this change does not have a large effect on module binary size. Apparently the order were were using before was already pretty good.

File: src/ir/module-utils.cpp [hunk 1]
```diff
@@ -684,95 +684,81 @@ std::vector<HeapType> getPrivateHeapTypes(Module& wasm) {
...
-      groupInfos.sort(sortedGroups);
-      for (auto group : sortedGroups) {
-        push(group);
-      }
+  // Each rec group's weight, adjusted for its size and incorporating the weight
+  // of its users.
+  std::vector<float> weights(groups.size());
+  for (size_t i = 0; i < groups.size(); ++i) {
+    weights[i] = (float)groupCounts[i] / groups[i].size();
+  }
```
Comment (kripken) [inline src/ir/module-utils.cpp:739]: Maybe another option than dividing by group size is to compute the benefit (which is nonlinear) as we go. Example to show what I mean:
```
group A {
  a // weight 101
}
group B {
  b1 // weight 190
  b2 // weight  10
}
```
The average of group B is 100, but if the first entry gets an index smaller than the two after it, then `b1,b2,a` will be smaller than `a,b1,b2`.

But maybe this is not worth the complexity.

**Review by kripken — APPROVED** [review]

File: src/ir/module-utils.cpp (hunk 1 above)
Comment (tlively) [reply src/ir/module-utils.cpp:739]: A tricky thing about this is that the rate of change of the benefit changes depending on the final indices the group will occupy, but maybe an approximation would be enough here.

I tried the related technique of approximating total cost of an order using the 1/128 smearing you used in ReorderGlobals, but it turned out to produce worse results than doing a more accurate count of the costs by calculating the exact byte size of each index.

Comment (kripken) [reply src/ir/module-utils.cpp:739]: Sounds good. Yeah, this is probably good enough as is.

**Comment by kripken** [comment]
Btw, what about ordering within a rec group, do we do that?

**Comment by tlively** [comment]
Not yet, since that changes type identity. I'm going to repurpose the removed-unused-types pass to do that, though.

**Pull request merged**